### PR TITLE
Add cosmos serverless spark batch job submission support

### PR DESCRIPTION
**Pull request by t-rufang**

### CODE REVIEW

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/spark/run/CosmosServerlessSparkBatchRunner.kt [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+    @Throws(ExecutionException::class)
+    override fun buildSparkBatchJob(submitModel: SparkSubmitModel, ctrlSubject: Observer<AbstractMap.SimpleImmutableEntry<MessageInfoType, String>>): ISparkBatchJob {
+        val submissionParameter = submitModel.submissionParameter as CreateSparkBatchJobParameters
+        val adlAccountName = submissionParameter.adlAccountName()
+        val account = AzureSparkServerlessClusterManager.getInstance().getAccountByName(adlAccountName)
+        account ?: throw ExecutionException("Can't find ADLA account '$adlAccountName'")
+
+        val accessToken = try {
+            account.http.accessToken
+        } catch (ex: IOException) {
```
Comment (wezhang) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/spark/run/CosmosServerlessSparkBatchRunner.kt]: Who throws IOException?

File: Utils/hdinsight-node-common/Test/java/com/microsoft/azure/hdinsight/spark/common/CosmosServerlessSparkBatchJobScenario.kt [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+        `when`(serverlessJobMock.log()).thenReturn(loggerMock)
+        doReturn(http).`when`(serverlessJobMock).http
+        // Actually submissionParameter is merely set to make serverlessJobMock.getSubmissionParameter() not break
+        // the real request body is defined with variable "json" below
+        doReturn(submissionParameter).`when`(serverlessJobMock).getSubmissionParameter()
+        doReturn(connectUri).`when`(serverlessJobMock).connectUri
+
+        // generate request json with submission parameter might lead to out-of-order in the json
+        // therefore, we need to define json here to make the real request body exactly same with the mock one
+        val json = """ {"name":"ServerlessSession_46c07889-3590-48f8-b2bc-7f52622b5a0b","priority":1000,"properties":{"className":"sample.JavaSparkPi","file":"adl://test.dogfood.net/rufan/serverless/artifact/serverless_artifact.jar","sparkEventsDirectoryPath":"adl://test.dogfood.net/spark-events","name":"Session_46c07889-3590-48f8-b2bc-7f52622b5a0b","driverMemory":"12G","driverCores":4,"executorCores":4,"numExecutors":2,"executorMemory":"12G"}}"""
```
Comment (wezhang) [inline Utils/hdinsight-node-common/Test/java/com/microsoft/azure/hdinsight/spark/common/CosmosServerlessSparkBatchJobScenario.kt]: Can the json be put into the feature file?

File: Utils/hdinsight-node-common/src/com/microsoft/azure/hdinsight/sdk/rest/azure/serverless/spark/models/SparkBatchJobResponsePayload.java [hunk 3]
```diff
@@ -23,90 +23,94 @@
 
 package com.microsoft.azure.hdinsight.sdk.rest.azure.serverless.spark.models;
 
-import java.util.Map;
-import java.util.List;
+//import java.util.Map;
+//import java.util.List;
```
Comment (wezhang) [inline Utils/hdinsight-node-common/src/com/microsoft/azure/hdinsight/sdk/rest/azure/serverless/spark/models/SparkBatchJobResponsePayload.java]: could be removed.

File: Utils/hdinsight-node-common/src/com/microsoft/azure/hdinsight/sdk/rest/azure/serverless/spark/models/SparkBatchJobResponsePayload.java [hunk 4]
```diff
@@ -23,90 +23,94 @@
...
 
 import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
-import com.fasterxml.jackson.annotation.JsonProperty;
+//import com.fasterxml.jackson.annotation.JsonProperty;
```
Comment (wezhang) [inline Utils/hdinsight-node-common/src/com/microsoft/azure/hdinsight/sdk/rest/azure/serverless/spark/models/SparkBatchJobResponsePayload.java]: could be removed.

File: Utils/hdinsight-node-common/src/com/microsoft/azure/hdinsight/spark/common/CosmosServerlessSparkBatchJob.java [hunk 5]
```diff
@@ -0,0 +1,235 @@
...
+        return http;
+    }
+
+    @NotNull
+    public CreateSparkBatchJobParameters getSubmissionParameter() {
+        return (CreateSparkBatchJobParameters) super.getSubmissionParameter();
+    }
+
+    // WARNING: This method should only be called for Unit Test
+    public void setJobId(@NotNull String jobId) {
```
Comment (wezhang) [inline Utils/hdinsight-node-common/src/com/microsoft/azure/hdinsight/spark/common/CosmosServerlessSparkBatchJob.java]: You can mock it in UT

File: Utils/hdinsight-node-common/src/com/microsoft/azure/hdinsight/spark/common/CosmosServerlessSparkBatchJob.java [hunk 6]
```diff
@@ -0,0 +1,235 @@
...
+    @NotNull
+    public CreateSparkBatchJob prepareCreateSparkBatchJob(CreateSparkBatchJobParameters parameters) {
+        return new CreateSparkBatchJob()
+                // TODO: Replace name with a meaningful name
+                .withName("ServerlessSession_" + jobId)
+                .withProperties(parameters);
+    }
+
+    @NotNull
+    public Observable<com.microsoft.azure.hdinsight.sdk.rest.azure.serverless.spark.models.SparkBatchJob> createSparkBatchJobRequest(@NotNull String jobId, @NotNull CreateSparkBatchJobParameters parameters) {
```
Comment (wezhang) [inline Utils/hdinsight-node-common/src/com/microsoft/azure/hdinsight/spark/common/CosmosServerlessSparkBatchJob.java]: Make a link break...

File: Utils/hdinsight-node-common/src/com/microsoft/azure/hdinsight/spark/common/CosmosServerlessSparkBatchJob.java [hunk 7]
```diff
@@ -0,0 +1,235 @@
...
+        String url = getConnectUri().toString() + "/" + jobId;
+
+        CreateSparkBatchJob putBody = prepareCreateSparkBatchJob(parameters);
+
+        String json = putBody.convertToJson()
+                .orElseThrow(() -> new IllegalArgumentException("Bad Spark Batch Job arguments to put"));
+
+        StringEntity entity = new StringEntity(json, StandardCharsets.UTF_8);
+        entity.setContentType("application/json");
```
Comment (wezhang) [inline Utils/hdinsight-node-common/src/com/microsoft/azure/hdinsight/spark/common/CosmosServerlessSparkBatchJob.java]: Already set in default headers

File: Utils/hdinsight-node-common/src/com/microsoft/azure/hdinsight/spark/common/CosmosServerlessSparkBatchJob.java [hunk 8]
```diff
@@ -0,0 +1,235 @@
...
+import java.util.AbstractMap;
+import java.util.UUID;
+import java.util.concurrent.TimeUnit;
+
+public class CosmosServerlessSparkBatchJob extends SparkBatchJob {
+    private static final String REST_SEGMENT_SPARK_BATCH_JOB = "/activityTypes/spark/batchJobs";
+    @NotNull
+    private final AzureSparkServerlessAccount account;
+    @NotNull
+    private String jobId;
```
Comment (wezhang) [inline Utils/hdinsight-node-common/src/com/microsoft/azure/hdinsight/spark/common/CosmosServerlessSparkBatchJob.java]: Could you change the `jobId` to `jobUuid`, which can be distinguished with Livy response spark job 'id'?

Comment (wezhang) [reply Utils/hdinsight-node-common/src/com/microsoft/azure/hdinsight/spark/common/CosmosServerlessSparkBatchJob.java]: Also include the following function parameters

File: Utils/hdinsight-node-common/src/com/microsoft/azure/hdinsight/spark/common/CosmosServerlessSparkBatchJob.java [hunk 9]
```diff
@@ -0,0 +1,235 @@
...
+        String url = getConnectUri().toString() + "/" + jobId;
+
+        CreateSparkBatchJob putBody = prepareCreateSparkBatchJob(parameters);
+
+        String json = putBody.convertToJson()
+                .orElseThrow(() -> new IllegalArgumentException("Bad Spark Batch Job arguments to put"));
+
+        StringEntity entity = new StringEntity(json, StandardCharsets.UTF_8);
+        entity.setContentType("application/json");
+
```
Comment (wezhang) [inline Utils/hdinsight-node-common/src/com/microsoft/azure/hdinsight/spark/common/CosmosServerlessSparkBatchJob.java]: Could you log the `PUT` and `DELETE` requests?

File: Utils/hdinsight-node-common/src/com/microsoft/azure/hdinsight/spark/common/CosmosServerlessSparkBatchJob.java [hunk 10]
```diff
@@ -0,0 +1,235 @@
...
+    @Nullable
+    @Override
+    public String getState() throws IOException {
+        return getStatus()
+                .retry(getRetriesMax())
+                .repeatWhen(ob -> ob.delay(getDelaySeconds(), TimeUnit.SECONDS))
+                .map(sparkSubmitResponse -> sparkSubmitResponse.getState())
+                .toBlocking()
+                .single();
```
Comment (wezhang) [inline Utils/hdinsight-node-common/src/com/microsoft/azure/hdinsight/spark/common/CosmosServerlessSparkBatchJob.java]: `singleOrDefault`?

File: Utils/hdinsight-node-common/src/com/microsoft/azure/hdinsight/spark/common/CosmosServerlessSparkBatchJob.java [hunk 11]
```diff
@@ -0,0 +1,235 @@
...
+    }
+
+    @Override
+    public boolean isActive() throws IOException {
+        return getStatus()
+                .retry(getRetriesMax())
+                .repeatWhen(ob -> ob.delay(getDelaySeconds(), TimeUnit.SECONDS))
+                .map(sparkSubmitResponse -> sparkSubmitResponse.isAlive())
+                .toBlocking()
+                .single();
```
Comment (wezhang) [inline Utils/hdinsight-node-common/src/com/microsoft/azure/hdinsight/spark/common/CosmosServerlessSparkBatchJob.java]: `singleOrDefault`?

File: Utils/hdinsight-node-common/src/com/microsoft/azure/hdinsight/spark/common/CosmosServerlessSparkBatchJob.java (hunk 9 above)
Comment (t-rufang) [reply Utils/hdinsight-node-common/src/com/microsoft/azure/hdinsight/spark/common/CosmosServerlessSparkBatchJob.java]: Won't fix in this PR

---
In reply to: [234894059](https://github.com/Microsoft/azure-tools-for-java/pull/2367#discussion_r234894059) [](ancestors = 234894059)

**Pull request merged**